model/util.go

```diff
@@ -17,114 +17,246 @@ package model
-	// #nosec
-	if resp, err := http.Get(s); err == nil {
+	return message
+}
+
+func getBytesFromFile(uri string) (b []byte, err error) {
+	refUrl, _ := url.Parse(uri)
+
+	if refUrl.Scheme == "" || refUrl.Scheme == "file" {
+		path := filepath.Join(refUrl.Host, refUrl.Path)
+		if !filepath.IsAbs(path) {
+			// The import file is an non-absolute path, we join it with include path
+			// TODO: if the file didn't find in any include path, we should report an error
+			for _, p := range IncludePaths() {
+				sn := filepath.Join(p, path)
+				if _, err := os.Stat(sn); err == nil {
+					path = sn
+					break
+				}
+			}
+		}
+
+		b, err = os.ReadFile(filepath.Clean(path))
+		if err != nil {
+			return nil, err
+		}
+
+	} else {
+		// #nosec
+		req, err := http.NewRequest(http.MethodGet, refUrl.String(), nil)
+		if err != nil {
+			return nil, err
+		}
+
+		resp, err := httpClient.Do(req)
+		if err != nil {
+			return nil, err
+		}
 		defer resp.Body.Close()
+
 		buf := new(bytes.Buffer)
 		if _, err = buf.ReadFrom(resp.Body); err != nil {
 			return nil, err
 		}
-		return buf.Bytes(), nil
-	}
-	s = strings.TrimPrefix(s, prefix)
-
-	if !filepath.IsAbs(s) {
-		// The import file is an non-absolute path, we join it with include path
-		// TODO: if the file didn't find in any include path, we should report an error
-		for _, p := range IncludePaths() {
-			sn := filepath.Join(p, s)
-			if _, err := os.Stat(sn); err == nil {
-				s = sn
-				break
-			}
-		}
-	}
 
-	if b, err = os.ReadFile(filepath.Clean(s)); err != nil {
-		return nil, err
+		b = buf.Bytes()
 	}
 
 	// TODO: optimize this
 	// NOTE: In specification, we can declare independent definitions with another file format, so
 	// we must convert independently yaml source to json format data before unmarshal.
-	if strings.HasSuffix(s, ".yaml") || strings.HasSuffix(s, ".yml") {
+	if !json.Valid(b) {
 		b, err = yaml.YAMLToJSON(b)
 		if err != nil {
 			return nil, err
 		}
 	}
+
 	return b, nil
 }
 
-func requiresNotNilOrEmpty(value interface{}) string {
-	if value == nil {
-		return ""
+func unmarshalObjectOrFile[U any](parameterName string, data []byte, valObject *U) error {
+	var valString string
+	err := unmarshalPrimitiveOrObject(parameterName, data, &valString, valObject)
+	if err != nil || valString == "" {
+		return err
+	}
+
+	// Assumes that the value inside `data` is a path to a known location.
+	// Returns the content of the file or a not nil error reference.
+	data, err = getBytesFromFile(valString)
+	if err != nil {
+		return err
 	}
-	return value.(string)
+
+	return unmarshalObject(parameterName, data, valObject)
 }
 
-// TODO: check the places that use unmarshalString if the case changes for primitiveOrStruct.
-func unmarshalString(data []byte) (string, error) {
-	var value string
-	if err := json.Unmarshal(data, &value); err != nil {
-		return "", err
+func unmarshalArrayOrFile[U any](parameterName string, data []byte, valObject *U) error {
```

Comment:
This function's implement is same as unmarshalObjectOrFile，can we just keep one of them？

Reply:
Keeping the function's name I think is important, but I did a refactoring, to avoid repetition.